### PR TITLE
WIP: instrument julia to get per-method-instance llvm optimization timings

### PR DESCRIPTION
**Pull request by NHDaly**
This is a WIP PR that adds Dtrace timings to julia to allow printing timing information per method instance.

This was pair-programmed with @vchuravy 😊

We're currently trying to measure the `llvm` optimization times, but long-term it'd be nice to do lots of different sections.
Also, `dtrace` might only work on macOS/solaris, but there's an equivalent tracing tool for linux that should work with the same configuration files.

### CODE REVIEW

**Comment by NHDaly** [comment]
This is the D script we're currently running to record the timings:

```D
julia*:::compile-start
{
  self->ts = timestamp;
}
julia*:::compile-end
/timestamp - self->ts > 1000/
{
  @timings[copyinstr(arg0)] = quantize((timestamp - self->ts)/1000);  // microseconds
}
```

File: src/codegen.cpp [hunk 1]
```diff
@@ -1080,6 +1108,9 @@ jl_code_instance_t *jl_compile_linfo(jl_method_instance_t *mi, jl_code_info_t *s
 
     // Codegen lock held in this block
     {
+        // Instrument the time spent in this block for this method instance
+        CompileTiming _c(mi);
```
Comment (JeffBezanson) [inline src/codegen.cpp]: This will not time LLVM optimizations; see the comment at the top of this function.

Comment (NHDaly) [reply src/codegen.cpp]: Ah, yeah, we just realized that as well.

I've pushed up new commits that move the timings to `jitlayers.cpp`. Now the trouble is trying to figure out _which_ method we're compiling! 😅

Comment (NHDaly) [reply src/codegen.cpp]: Sorry i hadn't pushed that until just now

File: src/jitlayers.cpp [hunk 2]
```diff
@@ -336,9 +339,21 @@ void JuliaOJIT::DebugObjectRegistrar::operator()(RTDyldObjHandleT H,
...
+  ~CompileTiming() {
+      JULIA_COMPILE_END(name.c_str());
+  }
+  std::string name;
+};
 
 object::OwningBinary<object::ObjectFile> JuliaOJIT::CompilerT::operator()(Module &M)
 {
+    // Instrument the time spent in this block for this method instance
+    CompileTiming _c(M.getName().str());
```
Comment (NHDaly) [inline src/jitlayers.cpp]: @JeffBezanson  do you know where I can find the caller of this function in a context that would have the name of the method instance?

Comment (NHDaly) [reply src/jitlayers.cpp]: Ah, okay, i put a breakpoint and got a stacktrace, and it gives me:

```
    frame #4: 0x00000001001b6ab7 libjulia.1.2.dylib`JuliaOJIT::addModule(this=<unavailable>, M=unique_ptr<llvm::Module, std::__1::default_delete<llvm::Module> > @ 0x00007ffeefbfe630) at jitlayers.cpp:477:41 [opt]
    frame #5: 0x00000001001b73a3 libjulia.1.2.dylib`jl_finalize_function(llvm::StringRef) [inlined] jl_add_to_ee(m=unique_ptr<llvm::Module, std::__1::default_delete<llvm::Module> > @ scalar) at jitlayers.cpp:702:25 [opt]
    frame #6: 0x00000001001b738f libjulia.1.2.dylib`jl_finalize_function(F=<unavailable>) at jitlayers.cpp:710 [opt]
    frame #7: 0x00000001001596e9 libjulia.1.2.dylib`getAddressForFunction(fname=<unavailable>) at codegen.cpp:1338:5 [opt]
    frame #8: 0x0000000100159a77 libjulia.1.2.dylib`::jl_generate_fptr(output=0x000000010c443c10) at codegen.cpp:1432:39 [opt]
    frame #9: 0x00000001000ddb01 libjulia.1.2.dylib`jl_compile_method_internal(mi=0x000000010c524350, world=25987) at gf.c:1795:5 [opt]
    frame #10: 0x00000001000e02ab libjulia.1.2.dylib`jl_apply_generic(args=0x00007ffeefbfe790, nargs=2) at gf.c:2196:16 [opt]
```

But i'm surprised; isn't `finalize` happening _after_ the optimization passes have finished? Why is the `JL_TIMING(LLVM_OPT);` block there?

**Comment by vchuravy** [comment]
I would love to have these kinds of instrumentation points, they are really useful for doing application specific probing and this just turned out to be an interesting test-bed to do this.
Right now this is Mac only, but in theory we should be able to use `bpftrace` on Linux as well.

**Comment by NHDaly** [comment]
Pasting the updated script I've been using to get a .csv of timings:

`test.d`
```D
#pragma D option strsize=10000

julia*:::compile-start
{
  self->ts = timestamp;
}
julia*:::compile-end
{
  @timings[copyinstr(arg0)] = max((timestamp - self->ts)/1000);  // microseconds
}
END
{
  printf("\n");
  printf("\n---- DTRACE ----\n");
  printa("%s\t%@d\n", @timings);
}
```
Then I run it like this, and delete everything up to the `---- DTRACE ----`:
```bash
$ sudo dtrace -s ~/src/julia/test.d -c '/Users/nathan.daly/src/julia/julia --project test-script.jl' > out.csv
```

**Comment by NHDaly** [comment]
The one remaining question I have before sending this for review is how to fix the Makefile.

Currently, modifying the `uprobes.d` file doesn't cause subsequent `make`s to rebuild. You can to do `cd src; make`.
Do I maybe also need to add `uprobes.d` to a rule in the toplevel `Makefile`?

**Comment by vchuravy** [comment]
For those following along at home I prototyped a version that can be used to create and trigger probes from Julia functions https://github.com/vchuravy/UProbes.jl

It's a tad bit more expensive than the C version (has to do a `ccall` instead of a `nop`). Maybe I can claw back that performance, but I am not to concerned with that right now.

**Pull request closed without merging**